### ISSUE TITLE
SFTP to an network device (Huawei VRP)

### ISSUE THREAD

**Issue by nielsvanhooy** [ronf/asyncssh#530]
Hi there.
i was building a wrapper around the SFTP functionality of this library when i ran into a problem.
im trying to get a file from a network device (huawei device). but i'm getting an error:

below my code:

            async with connect(**scp_options) as ssh_conn:
                if operation == "get":
                    async with ssh_conn.start_sftp_client() as sftp:
                        await sftp.get(
                            remotepaths="flash:/config",
                            localpath=".",
                        )

the device says its tftp version 3
sftp> version
SFTP protocol version 3
sftp> 

await sftp.get(
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 3793, in get
    await self._begin_copy(self, local_fs, remotepaths, localpath, 'get',
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 3661, in _begin_copy
    srcattrs = await srcfs.stat(srcpath)
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 4564, in stat
    return await self._handler.stat(path, flags)
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2686, in stat
    return cast(SFTPAttrs,  await self._make_request(
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2445, in _make_request
    result = self._packet_handlers[resptype](self, resp)
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2514, in _process_attrs
    attrs = SFTPAttrs().decode(packet, self._version)
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 1740, in decode
    raise SFTPBadMessage('Unsupported attribute flags: 0x%08x' %
asyncssh.sftp.SFTPBadMessage: Unsupported attribute flags: 0x00000020 <------------

when i turn on debugging on this router:

Dec 14 2022 14:57:14.1.1+01:00 huawei_device 10.1.1.138 SSH/7/RECV_PKT:Received SSH2_MSG_GLOBAL_REQUEST packet.
<huawei_device>
Dec 14 2022 14:57:14.1.2+01:00 huawei_device 10.1.1.138 SSH/7/GLOBALREQUEST:Receive GlobalRequest packet, type = keepalive@openssh.com, wantreplay = need reply.
<huawei_device>
Dec 14 2022 14:57:14.1.3+01:00 huawei_device 10.1.1.138 SSH/7/RECV_PKT:Received SSH2_MSG_GLOBAL_REQUEST packet.
<huawei_device>
Dec 14 2022 14:57:14.1.4+01:00 huawei_device 10.1.1.138 SSH/7/GLOBALREQUEST:Receive GlobalRequest packet, type = keepalive@openssh.com, wantreplay = need reply.
<huawei_device>
Dec 14 2022 14:57:14.1.5+01:00 huawei_device 10.1.1.138 SSH/7/SFTP_PROCESS:Sftp operation failed(client didn't send flags)! <------------------------------------------
<huawei_device>
Dec 14 2022 14:57:14.3.1+01:00 huawei_device 10.1.1.138 SSH/7/FSM_MOVE:FSM moved from SSH_Main_SSHProcess to SSH_Main_Disconnect.
<huawei_device>
Dec 14 2022 14:57:14.4.1+01:00 huawei_device 10.1.1.138 SSH/7/stdin_data:SSH_CMSG_USER message received on VTY 2
data: 

<huawei_device>
Dec 14 2022 14:57:14.6.1+01:00 huawei_device 10.1.1.138 SSH/7/FSM_MOVE:FSM moved from Disconnect to Disconnect,send ExitMsg..
<huawei_device>
Dec 14 2022 14:57:14.7.1+01:00 huawei_device 10.1.1.138 SSH/7/stdin_data:SSH_CMSG_USER message received on VTY 2
data: exit
<huawei_device>
Dec 14 2022 14:57:14.7.2+01:00 huawei_device 10.1.1.138 SSH/7/stdin_data:SSH_CMSG_USER message received on VTY 2
data: 

<huawei_device>
Dec 14 2022 14:57:14.7.3+01:00 huawei_device 10.1.1.138 SSH/7/FSM_MOVE:FSM moved from SSH_Main_SSHProcess to SSH_Main_Disconnect.
<huawei_device>
Dec 14 2022 14:57:15.2.1+01:00 huawei_device 10.1.1.138 SSH/7/FSM_MOVE:FSM moved from Disconnect to Disconnect,send ExitMsg..
<huawei_device>

either im doing something wrong. or the implementation of SFTP on this device is kinda wonky,
please let me know if you have any thoughts

**Comment by ronf** [comment]
Yeah - this looks like an error in the server. It is claiming to be version 3, but the attribute flags defined for version 3 (see https://filezilla-project.org/specs/draft-ietf-secsh-filexfer-02.txt) include only:

```
   	#define SSH_FILEXFER_ATTR_SIZE          0x00000001
   	#define SSH_FILEXFER_ATTR_UIDGID        0x00000002
   	#define SSH_FILEXFER_ATTR_PERMISSIONS   0x00000004
   	#define SSH_FILEXFER_ATTR_ACMODTIME     0x00000008
   	#define SSH_FILEXFER_ATTR_EXTENDED      0x80000000
```

The attribute flag 0x20 would only be defined by SFTP version 4, where ACMODTIME is split into separate access and modify times (as defined in https://filezilla-project.org/specs/draft-ietf-secsh-filexfer-03.txt):

```
   	#define SSH_FILEXFER_ATTR_SIZE            0x00000001
   	#define SSH_FILEXFER_ATTR_PERMISSIONS     0x00000004
   	#define SSH_FILEXFER_ATTR_ACCESSTIME      0x00000008
   	#define SSH_FILEXFER_ATTR_CREATETIME      0x00000010
   	#define SSH_FILEXFER_ATTR_MODIFYTIME      0x00000020
   	#define SSH_FILEXFER_ATTR_ACL             0x00000040
   	#define SSH_FILEXFER_ATTR_OWNERGROUP      0x00000080
   	#define SSH_FILEXFER_ATTR_EXTENDED        0x80000000
```

AsyncSSH supports SFTP version 4 and these new flags (plus other flags added in SFTP versions 5-6). However, AsyncSSH will only supported getting back these flags from a peer which negotiates version 4 or higher.

It's not simply a matter of ignoring the extra flag bits, as the flags which are set tells you what fields to expect after this flags field and how to interpret them. If the other implementation asks for version 3, that says that 0x08 means to read just a single UInt32 to use as both access and modify time, and it's not defined what additional data to read when the 0x20 bit is set. With SFTP version 4, though, the 0x08 means to read one UInt32 to use as the access time and then 0x20 would indicate reading a second UInt32 to use as the modify time. We'd have to see if the Huawei device sends separate access and modify times or not in this case, but without setting the version correctly, it's going to make it difficult for AsyncSSH to know how the rest of that packet should be interpreted.

Does it work if you set sftp_version to 4 when starting the AsyncSSH SFTP client?

**Comment by nielsvanhooy** [comment]
@ronf 

thanks for your quick response.
i set the sftp version to 4 but unfortunatly the same error.

  File "/home/hooij804/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-3bwyL76K-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 3793, in get
    await self._begin_copy(self, local_fs, remotepaths, localpath, 'get',
  File "/home/hooij804/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-3bwyL76K-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 3661, in _begin_copy
    srcattrs = await srcfs.stat(srcpath)
  File "/home/hooij804/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-3bwyL76K-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 4564, in stat
    return await self._handler.stat(path, flags)
  File "/home/hooij804/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-3bwyL76K-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2686, in stat
    return cast(SFTPAttrs,  await self._make_request(
  File "/home/hooij804/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-3bwyL76K-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2445, in _make_request
    result = self._packet_handlers[resptype](self, resp)
  File "/home/hooij804/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-3bwyL76K-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2514, in _process_attrs
    attrs = SFTPAttrs().decode(packet, self._version)
  File "/home/hooij804/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-3bwyL76K-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 1740, in decode
    raise SFTPBadMessage('Unsupported attribute flags: 0x%08x' %
asyncssh.sftp.SFTPBadMessage: Unsupported attribute flags: 0x00000020


device debugging:
Dec 14 2022 20:03:12.3.3+01:00 huawei_device 10.1.1.131 SSH/7/RECV_PKT:Received SSH2_MSG_CHANNEL_REQUEST packet.
<huawei_device>
Dec 14 2022 20:03:12.5.1+01:00 huawei_device 10.1.1.131 SSH/7/SFTP_PROCESS:Sftp operation failed(client didn't send flags)!
<huawei_device>
Dec 14 2022 20:03:12.6.1+01:00 huawei_device 10.1.1.131 SSH/7/FSM_MOVE:FSM moved from SSH_Main_SSHProcess to SSH_Main_Disconnect.
<huawei_device>
Dec 14 2022 20:03:12.6.2+01:00 huawei_device 10.1.1.131 SSH/7/stdin_data:SSH_CMSG_USER message received on VTY 2
data: 

<huawei_device>
Dec 14 2022 20:03:12.7.1+01:00 huawei_device 10.1.1.131 SSH/7/FSM_MOVE:FSM moved from Disconnect to Disconnect,send ExitMsg..
<huawei_device>
Dec 14 2022 20:03:12.8.1+01:00 huawei_device 10.1.1.131 SSH/7/stdin_data:SSH_CMSG_USER message received on VTY 2
data: exit
<huawei_device>
Dec 14 2022 20:03:12.8.2+01:00 huawei_device 10.1.1.131 SSH/7/stdin_data:SSH_CMSG_USER message received on VTY 2
data: 

<huawei_device>
Dec 14 2022 20:03:12.8.3+01:00 huawei_device 10.1.1.131 SSH/7/FSM_MOVE:FSM moved from SSH_Main_SSHProcess to SSH_Main_Disconnect.
<huawei_device>
Dec 14 2022 20:03:13.2.1+01:00 huawei_device 10.1.1.131 SSH/7/FSM_MOVE:FSM moved from SSH Process to Disconnect,send ExitMsg..
<huawei_device>
Dec 14 2022 20:03:13.4.1+01:00 huawei_device 10.1.1.131 SSH/7/FSM_MOVE:FSM moved from Disconnect to Disconnect,send ExitMsg..


its at least good to know that it seems to be huawei's implementation that is not "correct"
i dont think theres an "easy" way around this huawei implementation with the lib (let me know if you have any thoughts or otherwise we can close this issue).
i will contact huawei tomorrow as i am in a position to open tickets with them seeing as were deploying huawei's at customers. (i also know that in future firmware there implementing SCP).
thank you for your thorough response!

**Comment by ronf** [comment]
I had a feeling that changing to version 4 might not work. I'm guessing if you look more closely at the messages, you'll see the server as only claiming to support version 3, even when we offer a higher version, and despite the fact that it's returning a flags value that's not legal for version 3.

I might be able to work around things here, by looking to try and read a second UInt32 when this illegal flag is set, much like it would for version 4, and perhaps I could ignore this illegal flag when there are no remaining bytes, but this might not work if SSH_FILEXFER_ATTR_EXTENDED is set. If you could capture a debug log of message where you see the exception, that would be helpful (to see if there's an extra time value in it or not). Also, it would be helpful to capture the name which appears in the server version string, so this workaround could potentially be tied to that.

To get this info, just add the following lines to just before you start the AsyncSSH client:

```python
    import logging
    logging.basicConfig(level='DEBUG')
    asyncssh.set_debug_level(2)
```

The specific lines I'm looking for would be something like:

```
DEBUG:asyncssh:[conn=0] Sending version SSH-2.0-AsyncSSH_2.12.0
DEBUG:asyncssh:[conn=0] Received version SSH-2.0-OpenSSH_9.1
```

and then later:

```
INFO:asyncssh.sftp:[conn=0, chan=0] Starting SFTP client
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending init, version=3
DEBUG:asyncssh:[conn=0, chan=0] Sending 9 data bytes
DEBUG:asyncssh:[conn=0, chan=0] Reading from channel started
DEBUG:asyncssh:[conn=0, chan=0] Received 322 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received version=3, extensions:
DEBUG:asyncssh.sftp:[conn=0, chan=0]   posix-rename@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   statvfs@openssh.com: 2
DEBUG:asyncssh.sftp:[conn=0, chan=0]   fstatvfs@openssh.com: 2
DEBUG:asyncssh.sftp:[conn=0, chan=0]   hardlink@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   fsync@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   lsetstat@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   limits@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   expand-path@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   copy-data: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   home-directory: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   users-groups-by-id@openssh.com: 1
```

Also, look for where your get begins:

```
INFO:asyncssh.sftp:[conn=0, chan=0] Starting SFTP get of /tmp/data to /tmp/data1
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending stat for /tmp/data
DEBUG:asyncssh:[conn=0, chan=0] Sending 22 data bytes
DEBUG:asyncssh:[conn=0, chan=0] Received 41 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received type: regular, size: 4, uid: 501, gid: 0, permissions: 100644, atime: Wed Dec 14 18:54:40 2022, mtime: Wed Dec 14 18:54:40 2022
```

You probably won't see this last "Received" line before the exception is raised, but I'd be curious how many data bytes that last received message is before the exception.

**Comment by nielsvanhooy** [comment]
@ronf 
DEBUG:asyncssh:[conn=2] Sending version SSH-2.0-AsyncSSH_2.12.0
DEBUG:asyncssh:[conn=2] Received version SSH-2.0-HUAWEI-1.5


and for the SFTP:

INFO:asyncssh:[conn=2, chan=0] Requesting new SSH session
DEBUG:asyncssh:[conn=2, chan=0]   Initial recv window 2097152, packet size 32768
DEBUG:asyncssh:[conn=2, chan=0]   Initial send window 131072, packet size 32768
INFO:asyncssh:[conn=2, chan=0]   Subsystem: sftp
INFO:asyncssh.sftp:[conn=2, chan=0] Starting SFTP client
DEBUG:asyncssh.sftp:[conn=2, chan=0] Sending init, version=3
DEBUG:asyncssh:[conn=2, chan=0] Sending 9 data bytes
DEBUG:asyncssh:[conn=2, chan=0] Reading from channel started
DEBUG:asyncssh:[conn=2, chan=0] Received 9 data bytes
DEBUG:asyncssh.sftp:[conn=2, chan=0] Received version=3
INFO:asyncssh.sftp:[conn=2, chan=0] Starting SFTP get of flash:/21500104832SL5600370-1668688866694.cfg to 
DEBUG:asyncssh.sftp:[conn=2, chan=0] Sending stat for flash:/21500104832SL5600370-1668688866694.cfg
DEBUG:asyncssh:[conn=2, chan=0] Sending 58 data bytes
DEBUG:asyncssh:[conn=2, chan=0] Received 33 data bytes
DEBUG:asyncssh:[conn=2, chan=0] Sending EOF
INFO:asyncssh:[conn=2, chan=0] Received channel close
INFO:asyncssh:[conn=2, chan=0] Channel closed
INFO:asyncssh.sftp:[conn=2, chan=0] SFTP client exited
INFO:asyncssh:[conn=2, chan=0] Closing channel
INFO:asyncssh:[conn=2] Closing connection
INFO:asyncssh:[conn=2] Sending disconnect: Disconnected by application (11)
INFO:asyncssh:[conn=2] Connection closed

This should be the info you need.

**Comment by ronf** [comment]
Thanks!

If I had to guess, I'd say the 33 byte response probably has a flags field of 0x0000002b, which would translate into:

    pkttype (1 byte, 0x69)
    pktid (4 bytes)
    flags (4 bytes 0x0000002b?)
    size (8 bytes)
    uid (4 bytes)
    gid (4 bytes)
    atime (4 bytes)
    mtime (4 bytes)

This corresponds to everything but permissions (4 bytes) being encoded as attributes.

Thinking about this some more, it should be safe to ignore FILEXFER_ATTR_MODIFYTIME (0x20) when the SFTP version is 3 and FILEXFER_ATTR_ACMODTIME (0x08) is set, as either way that would actually involve sending both atime & mtime in the response in the same position relative to the other possible fields. As such, you might want to try the following patch:

```patch
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 5cc8498..bc5efc6 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -1735,6 +1735,10 @@ class SFTPAttrs(Record):
         flags = packet.get_uint32()
         attrs = cls()

+        if sftp_version == 3 and flags & (FILEXFER_ATTR_ACMODTIME |
+                                          FILEXFER_ATTR_MODIFYTIME):
+            flags &= ~FILEXFER_ATTR_MODIFYTIME
+
         unsupported_attrs = flags & ~_valid_attr_flags[sftp_version]

         if unsupported_attrs:
```

I can't test it here, but this may work around your issue. It'll clear the bad FILEXFER_ATTR_MODIFYTIME flag from being set before going into the check for unsupported attributes. Then, either way, it'll extract both the atime and mtime attributes as long as  FILEXFER_ATTR_ACMODTIME is set.

**Comment by nielsvanhooy** [comment]
@ronf.

that got it a bit further.


DEBUG:asyncssh:[conn=2] Sending version SSH-2.0-AsyncSSH_2.12.0
DEBUG:asyncssh:[conn=2] Received version SSH-2.0-HUAWEI-1.5


INFO:asyncssh:[conn=2, chan=0]   Subsystem: sftp
INFO:asyncssh.sftp:[conn=2, chan=0] Starting SFTP client
DEBUG:asyncssh.sftp:[conn=2, chan=0] Sending init, version=3
DEBUG:asyncssh:[conn=2, chan=0] Sending 9 data bytes
DEBUG:asyncssh:[conn=2, chan=0] Reading from channel started
DEBUG:asyncssh:[conn=2, chan=0] Received 9 data bytes
DEBUG:asyncssh.sftp:[conn=2, chan=0] Received version=3
INFO:asyncssh.sftp:[conn=2, chan=0] Starting SFTP get of flash:/21500104832SL5600370-1668688866694.cfg to 
DEBUG:asyncssh.sftp:[conn=2, chan=0] Sending stat for flash:/21500104832SL5600370-1668688866694.cfg
DEBUG:asyncssh:[conn=2, chan=0] Sending 58 data bytes
DEBUG:asyncssh:[conn=2, chan=0] Received 33 data bytes
DEBUG:asyncssh.sftp:[conn=2, chan=0] Received type: regular, size: 13092, permissions: 100777, atime: Sun Jan 11 00:51:58 1970, mtime: Sun Jan 11 00:51:58 1970
INFO:asyncssh.sftp:[conn=2, chan=0]   Copying file flash:/21500104832SL5600370-1668688866694.cfg to 
DEBUG:asyncssh.sftp:[conn=2, chan=0] Sending open for flash:/21500104832SL5600370-1668688866694.cfg, mode 0x01
DEBUG:asyncssh:[conn=2, chan=0] Sending 66 data bytes
DEBUG:asyncssh:[conn=2, chan=0] Received 13 data bytes
DEBUG:asyncssh:[conn=2, chan=0] Sending EOF
INFO:asyncssh:[conn=2, chan=0] Received channel close
INFO:asyncssh:[conn=2, chan=0] Channel closed
INFO:asyncssh.sftp:[conn=2, chan=0] SFTP client exited
INFO:asyncssh:[conn=2, chan=0] Closing channel
INFO:asyncssh:[conn=2] Closing connection
INFO:asyncssh:[conn=2] Sending disconnect: Disconnected by application (11)
INFO:asyncssh:[conn=2] Connection closed

  File "/home/donnyio/git/scrapli_transfer_utils/scrapli_transfer_utils/asyncsftp/base.py", line 204, in file_transfer
    raise e
  File "/home/donnyio/git/scrapli_transfer_utils/scrapli_transfer_utils/asyncsftp/base.py", line 197, in file_transfer
    _transferred = await self._async_file_transfer(
  File "/home/donnyio/git/scrapli_transfer_utils/scrapli_transfer_utils/asyncsftp/base.py", line 137, in _async_file_transfer
    raise e
  File "/home/donnyio/git/scrapli_transfer_utils/scrapli_transfer_utils/asyncsftp/base.py", line 130, in _async_file_transfer
    await sftp.get(
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 3798, in get
    await self._begin_copy(self, local_fs, remotepaths, localpath, 'get',
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 3694, in _begin_copy
    await self._copy(srcfs, dstfs, srcfile, dstfile, srcname.attrs,
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 3613, in _copy
    await _SFTPFileCopier(block_size, max_requests, 0,
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 776, in run
    self._src = await self._srcfs.open(self._srcpath, 'rb')
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 4445, in open
    handle = await self._handler.open(path, pflags, attrs)
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2633, in open
    return cast(bytes, await self._make_request(
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2450, in _make_request
    result = self._packet_handlers[resptype](self, resp)
  File "/home/donnyio/.cache/pypoetry/virtualenvs/scrapli-transfer-utils-O6q87ddK-py3.10/lib/python3.10/site-packages/asyncssh/sftp.py", line 2466, in _process_status
    raise exc
asyncssh.sftp.SFTPFailure


it now gets up to the part of opening the file.
manual sftp from the server gives these responses on the device:

NORMAL SFTP usage in server
Dec 16 2022 09:37:19+01:00 one-gv-2595bk-03 10.1.1.131 %%01FTPS/5/SFTPS_REQUEST(l)[14]:The sftp user had a request. (UserName="username", IpAddress=192.168.28.197,  Request=open file flash:/21500104832SL5600370-1670576590861.cfg)
<one-gv-2595bk-03>
Dec 16 2022 09:37:19+01:00 one-gv-2595bk-03 10.1.1.131 %%01FTPS/5/SFTPS_REQUEST(l)[15]:The sftp user had a request. (UserName="username", IpAddress=192.168.28.197,  Request=get file flash:/21500104832SL5600370-1670576590861.cfg)


with asyncssh sftp
Dec 16 2022 09:47:01+01:00 one-gv-2595bk-03 10.1.1.131 %%01FTPS/5/SFTPS_REQUEST(l)[25]:The sftp user had a request. (UserName="username", IpAddress=192.168.28.197,  Request=open file flash:/21500104832SL5600370-1668688866694.cfg)
<one-gv-2595bk-03>

**Comment by ronf** [comment]
The filename appears to be different here. In the "normal" output, the filename is "flash:/21500104832SL5600370-1670576590861.cfg". However, in the AsyncSSH output, it appears to be "flash:/21500104832SL5600370-1670576590861", missing the ".cfg" at the end.

**Comment by ronf** [comment]
Looking at the AsyncSSH debug output, the ".cfg" appears to be present, but that's for a different path: "flash:/21500104832SL5600370-1668688866694.cfg".

**Comment by nielsvanhooy** [comment]
@ronf. youre right. i messed up my error example a bit with copy/pasting. and using different filenames.
i. have updated the example now in my previous comment.

**Comment by ronf** [comment]
There really isn't much to an SFTP "open" request. The only arguments are the filename, the mode (which is just FXP_READ in this case) and an SFTP attrs (which is empty when reading files). So, unless there's something wrong with the filename being sent by AsyncSSH, I can't really see how that specific request could fail on AsyncSSH but not on OpenSSH.

There is an extra "stat" call just before the open, and I have seen cases where an object that's designed to be read only once can return a failure if it is accessed a second time, but so far I've only see that when doing an fstat() on an already open file and not a stat() on a pathname.

Would it be possible get verbose debug output from the OpenSSH SFTP client, to see if anything is different in the sequence of SFTP requests it is sending?

**Comment by ronf** [comment]
I know there are still open issues here, but the fix for the first issue with the illegal flags  is now available in AsyncSSH 2.13.0.

**Comment by ronf** [comment]
Closing due to inactivity, but feel free to re-open this or create a new issue if you have additional information.
